compiler/rustc_builtin_macros/src/deriving/default.rs

```diff
@@ -15,6 +19,17 @@ pub fn expand_deriving_default(
     item: &Annotatable,
     push: &mut dyn FnMut(Annotatable),
 ) {
+    // Catch outer `#[default]` attribute on items.
+    if let Annotatable::Item(item) = item {
+        if let Some(attr) = item.attrs.iter().find(|attr| attr.has_name(kw::Default)) {
+            cx.struct_span_err(
+                attr.span,
+                "the `#[default]` attribute may only be used on unit enum variants",
+            )
+            .emit();
+        }
+    }
```

Comment:
This check and the struct field check are not enough.

You'll have to implement a small `Visitor` (`compiler\rustc_ast\src\visit.rs`), use it for walking the whole AST fragment passed to `Default`, and report this error from `fn visit_attribute` with exception for variants (can be done by overriding `fn visit_variant`).

Test case:
```rust
#[derive(Default)]
enum E {
    V = #[default] 0,
    U([u8; #[default] 1]),
}
```

Reply:
Didn't even know that attributes were syntactically permitted in that position.

Reply:
Ah, they're not on stable. It's attributes on expressions. Going ahead with implementing a visitor.